src/Data/RFC5322/Internal.hs

```diff
@@ -22,38 +46,210 @@ module Data.RFC5322.Internal
 
   ) where
 
-import Control.Applicative ((<|>), liftA2, many)
+import Prelude hiding (takeWhile)
+import Control.Applicative ((<|>), Alternative, liftA2, many, optional)
 import Control.Monad (void)
-import Data.Attoparsec.ByteString as A
-import Data.Attoparsec.Internal as A
+import Control.Lens.Cons (Cons, cons)
+import qualified Data.Attoparsec.ByteString as A
+import qualified Data.Attoparsec.Internal as A
 import qualified Data.Attoparsec.Internal.Types as AT
+import qualified Data.Attoparsec.Text as AText
 import qualified Data.ByteString as B
+import Data.ByteString.Internal (c2w, w2c)
 import Data.ByteString.Search (indices)
 import Data.CaseInsensitive (CI, FoldCase, mk, original)
+import Data.Char (chr)
 import Data.Foldable (fold)
 import Data.Functor (($>))
 import Data.List.NonEmpty (fromList)
 import Data.Semigroup (Semigroup((<>)))
 import Data.Semigroup.Foldable (fold1)
+import qualified Data.Text as T
+import Data.Word (Word8)
+
+
+-- | Constraint synonym to handle the Semigroup Monoid Proposal
+-- transition gracefully.
+#if MIN_VERSION_base(4,11,0)
+type SM a = Monoid a
+#else
+type SM a = (Semigroup a, Monoid a)
```

Comment:
`-XConstraintKinds` allows us to create "constraint synonyms" that we can then use on RHS of `=>`.
Since GHC 8.4, `Semigroup` is a superclass of `Monoid` but for earlier versions we need to explicit refer to both
classes.